Make a HEAD /foo requests emit metrics for HEAD:/foo, not GET:/foo

mikaelgoldmann opened the pull request:
This fixes an issue that probably originates from the fact that
apollo registers a HEAD endpoint automatically for GET endpoints.
This means a HEAD request is served by a GET handler and endpoint
metrics are incorrectly reported for endpoint = GET:/foo rather than
HEAD:/foo.

mikaelgoldmann committed:
Make a HEAD /foo requests emit metrics for HEAD:/foo, not GET:/foo

This fixes an issue that probably originates from the fact that
apollo registers a HEAD endpoint automatically for GET endpoints.
This means a HEAD request is served by a GET handler and endpoint
metrics are incorrectly reported for endpoint = GET:/foo rather than
HEAD:/foo.

rydenius commented on modules/metrics/src/main/java/com/spotify/apollo/metrics/MetricsCollectingEndpointRunnableFactoryDecorator.java:
```diff
@@ -42,7 +42,20 @@
+      } else {
+        // but because GET endpoints serve HEAD requests info().getName() is
+        // wrong does not distinguish HEAD and GET requests
+        endpointName = request.request().method() + ":" + endpoint.info().getUri();
```
Looks like this could work. But it would be more elegant if proper endpoint infos were set up for the implicit HEAD endpoints. This is a bit of a hack, isn't it?

mikaelgoldmann replied:
Definitely a hack. I'll see if I can come up with a better option.

mikaelgoldmann committed:
Simplifying the PR

String concatenation happens in the old code as well so doing it
when emitting the metric is not adding more string creation.

mikaelgoldmann commented:
I haven't had time for this so I will close it for now.
